src/os/bluestore/BlueStore.h

```diff
@@ -378,56 +380,77 @@ class BlueStore : public ObjectStore,
       ceph_assert(writing.empty());
     }
 
-    void _add_buffer(BufferCacheShard* cache, Buffer* b, int level, Buffer* near) {
+    void _add_buffer(BufferCacheShard *cache, BufferSpace *space, unsigned s,
```

Comment:
May be better to use ctor with move semantics for Buffer instead to avoid BufferSpace interface growth?
See attached patch for possible implementation.

Reply:
diff --git a/src/os/bluestore/BlueStore.cc b/src/os/bluestore/BlueStore.cc
index e6e1aa0541f..ed8eb34fd92 100644
--- a/src/os/bluestore/BlueStore.cc
+++ b/src/os/bluestore/BlueStore.cc
@@ -1728,9 +1728,9 @@ int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint32_t offset, u
 	if (b->data.length()) {
 	  bufferlist bl;
 	  bl.substr_of(b->data, b->length - tail, tail);
-	  _add_buffer(cache, this, b->state, b->seq, end, bl, b->flags, 0, 0, b);
+	  _add_buffer(cache, std::move(Buffer(this, b->state, b->seq, end, bl, b->flags)), 0, 0, b);
 	} else {
-	  _add_buffer(cache, this, b->state, b->seq, end, tail, b->flags, 0, 0, b);
+	  _add_buffer(cache, std::move(Buffer(this, b->state, b->seq, end, tail, b->flags)), 0, 0, b);
 	}
 	if (!b->is_writing()) {
 	  cache->_adjust_size(b, front - (int64_t)b->length);
@@ -1760,9 +1760,13 @@ int BlueStore::BufferSpace::_discard(BufferCacheShard* cache, uint32_t offset, u
     if (b->data.length()) {
       bufferlist bl;
       bl.substr_of(b->data, b->length - keep, keep);
-      _add_buffer(cache, this, b->state, b->seq, end, bl, b->flags, 0, 0, b);
+      _add_buffer(cache,
+                  std::move(Buffer(this, b->state, b->seq, end, bl, b->flags)),
+                  0, 0, b);
     } else {
-      _add_buffer(cache, this, b->state, b->seq, end, keep, b->flags, 0, 0, b);
+      _add_buffer(cache,
+                  std::move(Buffer(this, b->state, b->seq, end, keep, b->flags)),
+                  0, 0, b);
     }
     _rm_buffer(cache, i);
     cache->_audit("discard end 2");
@@ -1779,7 +1783,6 @@ void BlueStore::BufferSpace::read(
   interval_set<uint32_t>& res_intervals,
   int flags)
 {
-  auto start = std::chrono::high_resolution_clock::now();
   res.clear();
   res_intervals.clear();
   uint32_t want_bytes = length;
@@ -1842,11 +1845,6 @@ void BlueStore::BufferSpace::read(
   uint64_t miss_bytes = want_bytes - hit_bytes;
   cache->logger->inc(l_bluestore_buffer_hit_bytes, hit_bytes);
   cache->logger->inc(l_bluestore_buffer_miss_bytes, miss_bytes);
-  auto end_time = std::chrono::high_resolution_clock::now();
-  auto duration =
-      std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start);
-  std::cout << "Time taken: " << duration.count() << " ns"
-            << std::endl;
 }
 
 void BlueStore::BufferSpace::_finish_write(BufferCacheShard* cache, uint64_t seq)
@@ -1895,7 +1893,10 @@ bool BlueStore::BufferSpace::_dup_writing(BufferCacheShard* cache, BufferSpace*
     for (auto it = writing.begin(); it != writing.end(); ++it) {
       Buffer& b = *it;
       ceph_assert(b.is_writing());
-      to->_add_buffer(cache, to, b.state, b.seq, b.offset, b.data, b.flags, 0, 0, nullptr);
+      to->_add_buffer(cache,
+                      std::move(
+                        Buffer(to, b.state, b.seq, b.offset, b.data, b.flags)),
+                      0, 0, nullptr);
     }
   }
   return copied;
@@ -1918,11 +1919,15 @@ void BlueStore::BufferSpace::split(BufferCacheShard* cache, size_t pos, BlueStor
       if (p->second.data.length()) {
         bufferlist bl;
         bl.substr_of(p->second.data, left, right);
-        r._add_buffer(cache, &r, p->second.state, p->second.seq, 0, bl,
-                      p->second.flags, 0, 0, &p->second);
+        r._add_buffer(cache,
+                      std::move(Buffer(&r, p->second.state, p->second.seq,
+                                0, bl, p->second.flags)),
+                      0, 0, &p->second);
       } else {
-        r._add_buffer(cache, &r, p->second.state, p->second.seq, 0, right,
-                      p->second.flags, 0, 0, &p->second);
+        r._add_buffer(cache,
+                      std::move(Buffer(&r, p->second.state, p->second.seq,
+                                       0, right, p->second.flags)),
+                      0, 0, &p->second);
       }
       cache->_adjust_size(&p->second, -right);
       p->second.truncate(left);
@@ -1932,12 +1937,16 @@ void BlueStore::BufferSpace::split(BufferCacheShard* cache, size_t pos, BlueStor
     ceph_assert(p->second.end() > pos);
     ldout(cache->cct, 30) << __func__ << " move " << p->second << dendl;
     if (p->second.data.length()) {
-      r._add_buffer(cache, &r, p->second.state, p->second.seq,
-                    p->second.offset - pos, p->second.data, p->second.flags, 0,
-                    0, &p->second);
+      r._add_buffer(cache, std::move(
+                      Buffer(&r, p->second.state, p->second.seq,
+                             p->second.offset - pos, p->second.data,
+                             p->second.flags)),
+                    0, 0, &p->second);
     } else {
-      r._add_buffer(cache, &r, p->second.state, p->second.seq,
-                    p->second.offset - pos, p->second.length, p->second.flags,
+      r._add_buffer(cache, std::move(
+                      Buffer(&r, p->second.state, p->second.seq,
+                             p->second.offset - pos, p->second.length,
+                             p->second.flags)),
                     0, 0, &p->second);
     }
     if (p == buffer_map.begin()) {
@@ -2871,7 +2880,7 @@ uint32_t BlueStore::Blob::merge_blob(CephContext* cct, Blob* blob_to_dissolve)
     auto buf = src->bc.buffer_map.extract(src->bc.buffer_map.cbegin());
     buf.mapped().space = &dst->bc;
     if (dst->bc.buffer_map.count(buf.key()) == 0) {
-      dst->bc.buffer_map.insert({buf.key(), std::move(buf.mapped())});
+      dst->bc.buffer_map.emplace(std::make_pair(buf.key(), std::move(buf.mapped())));
     }
   }
   // move BufferSpace writing
diff --git a/src/os/bluestore/BlueStore.h b/src/os/bluestore/BlueStore.h
index d8c2893e0c1..b1cdf555cd1 100644
--- a/src/os/bluestore/BlueStore.h
+++ b/src/os/bluestore/BlueStore.h
@@ -304,6 +304,19 @@ public:
     boost::intrusive::list_member_hook<> lru_item;
     boost::intrusive::list_member_hook<> state_item;
 
+    Buffer(Buffer&& other) {
+      std::swap(space, other.space);
+      std::swap(state, other.state);
+      std::swap(cache_private, other.cache_private);
+      std::swap(flags, other.flags);
+      std::swap(seq, other.seq);
+      std::swap(offset, other.offset);
+      std::swap(length, other.length);
+      std::swap(data, other.data);
+      std::swap(cache_age_bin, other.cache_age_bin);
+      std::swap(lru_item, other.lru_item);
+      std::swap(state_item, other.state_item);
+    }
     Buffer(BufferSpace *space, unsigned s, uint64_t q, uint32_t o, uint32_t l,
 	   unsigned f = 0)
       : space(space), state(s), flags(f), seq(q), offset(o), length(l) {}
@@ -380,31 +393,13 @@ public:
       ceph_assert(writing.empty());
     }
 
-    void _add_buffer(BufferCacheShard *cache, BufferSpace *space, unsigned s,
-                     uint64_t q, uint32_t o, uint16_t length, unsigned f,
-                     uint16_t cache_private, int level, Buffer *near) {
-      auto it =
-          buffer_map.emplace(std::piecewise_construct, std::forward_as_tuple(o),
-                             std::forward_as_tuple(space, s, q, o, length, f));
-      Buffer *buffer = &it.first->second;
-      buffer->cache_private = cache_private;
-      _add_buffer(cache, space, buffer, level, near);
-    }
+    void _add_buffer(BufferCacheShard *cache, Buffer &&b, uint16_t cache_private, int level, Buffer *near) {
+      cache->_audit("_add_buffer start");
+      auto it = buffer_map.emplace(
+        std::make_pair(b.offset, std::move(b))); // b is invalidated after this point
 
-    void _add_buffer(BufferCacheShard *cache, BufferSpace *space, unsigned s,
-                     uint64_t q, uint32_t o, ceph::buffer::list &b, unsigned f,
-                     uint16_t cache_private, int level, Buffer *near) {
-      auto it =
-          buffer_map.emplace(std::piecewise_construct, std::forward_as_tuple(o),
-                             std::forward_as_tuple(space, s, q, o, b, f));
       Buffer *buffer = &it.first->second;
       buffer->cache_private = cache_private;
-      _add_buffer(cache, space, buffer, level, near);
-    }
-
-    __attribute__((always_inline)) void _add_buffer(BufferCacheShard *cache, BufferSpace *space,
-                     Buffer *buffer, int level, Buffer *near) {
-      cache->_audit("_add_buffer start");
       if (buffer->is_writing()) {
         // we might get already cached data for which resetting mempool is inppropriate
         // hence calling try_assign_to_mempool
@@ -471,18 +466,19 @@ public:
     void write(BufferCacheShard* cache, uint64_t seq, uint32_t offset, ceph::buffer::list& bl,
 	       unsigned flags) {
       std::lock_guard l(cache->lock);
-      Buffer b(this, Buffer::STATE_WRITING, seq, offset, bl,
-			     flags);
       uint16_t cache_private = _discard(cache, offset, bl.length());
-      _add_buffer(cache, this, Buffer::STATE_WRITING, seq, offset, bl, flags, cache_private, (flags & Buffer::FLAG_NOCACHE) ? 0 : 1, nullptr);
+      _add_buffer(cache, std::move(
+                    Buffer(this, Buffer::STATE_WRITING, seq, offset, bl, flags)),
+                  cache_private, !(flags & Buffer::FLAG_NOCACHE), nullptr);
       cache->_trim();
     }
     void _finish_write(BufferCacheShard* cache, uint64_t seq);
     void did_read(BufferCacheShard* cache, uint32_t offset, ceph::buffer::list& bl) {
       std::lock_guard l(cache->lock);
-      Buffer b(this, Buffer::STATE_CLEAN, 0, offset, bl);
       uint16_t cache_private = _discard(cache, offset, bl.length());
-      _add_buffer(cache, this, Buffer::STATE_CLEAN, 0, offset, bl, 0, cache_private, 1, nullptr);
+      _add_buffer(cache, std::move(
+                    Buffer(this, Buffer::STATE_CLEAN, 0, offset, bl)),
+                  cache_private, 1, nullptr);
       cache->_trim();
     }

Reply:
I was playing around https://quick-bench.com/ to understand what would happen with `piecewise_construct` or `_add_buffer(std::move(Buffer(...)));` and I had surprising results:

![RlsrIac1wT_YFQhEgI2eTsmFjro](https://github.com/ceph/ceph/assets/30913090/7769f163-2f03-4181-a7c9-b99d88516af5)

(y axis is ms, higher is slower)

As you can see what I did with piecewise_construct is surprisingly slow compared to just your patch. In addition to move, a reference and then insert instead of emplace is just as fast in this small bench.

[benchmark code](https://godbolt.org/z/szz8j8afc)

In conclusion, I will undoubtedly bring your patch alive

Reply:
Nice tool, wasn't aware of it.
I played a bit and got less optimistic view for insert() when using gcc 13.x on x86 and c++ 20.
![movector](https://github.com/ceph/ceph/assets/7138176/0257b7cd-29b9-4c6d-8c66-ccc19ef6fc1e)

Reply:
![OaHvA52utZs6z4cArgPLwk2fN0Y](https://github.com/ceph/ceph/assets/30913090/38dd730e-0bca-4788-89c5-d82600d94a74)

I saw a warning in the compiler that `std::move(Buffer(...))` prevents copy elision optimization so I think I'll go without std::move

Reply:
> Nice tool, wasn't aware of it.

Same!  That's pretty cool!